crates/oxc_parser/src/lexer/simd.rs

```diff
@@ -39,7 +41,7 @@ impl SkipWhitespace {
     }
 
     pub fn simd(mut self, bytes: &[u8]) -> Self {
-        let (chunks, remainder) = bytes.as_chunks::<ELEMENTS>();
+        let (chunks, _remainder) = bytes.as_chunks::<ELEMENTS>();
```

Comment:
Don't we still pay for `as_chunks`'s bounds checking?

Reply:
As far as I can tell there's a single bound check in `split_at`, but I don't think that's what you were referring to?
```rust
pub fn as_chunks<const N: usize>(&self) -> (&[[T; N]], &[T]) {
        assert_ne!(N, 0, "chunks cannot have a size of zero");
        let len = self.len() / N;
        let (multiple_of_n, remainder) = self.split_at(len * N);
        // SAFETY: We already panicked for zero, and ensured by construction
        // that the length of the subslice is a multiple of N.
        let array_slice = unsafe { multiple_of_n.as_chunks_unchecked() };
        (array_slice, remainder)
    }

pub const fn split_at(&self, mid: usize) -> (&[T], &[T]) {
        assert!(mid <= self.len());
        // SAFETY: `[ptr; mid]` and `[mid; len]` are inside `self`, which
        // fulfills the requirements of `split_at_unchecked`.
        unsafe { self.split_at_unchecked(mid) }
    }

pub unsafe fn as_chunks_unchecked<const N: usize>(&self) -> &[[T; N]] {
        let this = self;
        // SAFETY: Caller must guarantee that `N` is nonzero and exactly divides the slice length
        let new_len = unsafe {
            assert_unsafe_precondition!(
                "slice::as_chunks_unchecked requires `N != 0` and the slice to split exactly into `N`-element chunks",
                [T](this: &[T], N: usize) => N != 0 && this.len() % N == 0
            );
            exact_div(self.len(), N)
        };
        // SAFETY: We cast a slice of `new_len * N` elements into
        // a slice of `new_len` many `N` elements chunks.
        unsafe { from_raw_parts(self.as_ptr().cast(), new_len) }
    }
```

I got curious to look at the assembly (not something I'm well versed in). 
```asm
.LBB2329_1:
     cmp     r9, r10
     je      .LBB2329_12
```
is the single comparison (to see if we've reached the end of the array/chunks) for each iteration.
Isn't that unavoidable?
(I'll admit I still haven't understood why it works together with the start of the function `and     r9, -16` which zeroes the 16 LSBs)

`cargo asm --asm-style=intel --rust oxc_parser::lexer::simd::SkipWhitespace::simd` (after sprinkling some `pub use` so it's visible):
```asm
pub fn simd(mut self, bytes: &[u8]) -> Self {
 push    rsi
 push    rdi
 push    rbx
 mov     rax, rcx
 movzx   ecx, byte, ptr, [rdx, +, 73]
 for chunk in chunks {
 and     r9, -16
 xor     r10d, r10d
 mov     r11d, 65535
.LBB2329_1:
     cmp     r9, r10
     je      .LBB2329_12
     movdqu  xmm1, xmmword, ptr, [r8, +, r10]
     movdqa  xmm2, xmmword, ptr, [rdx]
     pcmpeqb xmm2, xmm1
     movdqa  xmm0, xmmword, ptr, [rdx, +, 16]
     pcmpeqb xmm0, xmm1
     por     xmm0, xmm2
     movdqa  xmm2, xmmword, ptr, [rdx, +, 32]
     pcmpeqb xmm2, xmm1
     pcmpeqb xmm1, xmmword, ptr, [rdx, +, 48]
     por     xmm1, xmm2
     por     xmm1, xmm0
     pmovmskb esi, xmm1
 let advance_by = (!any_white.to_bitmask()).trailing_zeros();
 not     esi
     test    si, si
     je      .LBB2329_3
     bsf     si, si
 if !self.newline
 test    cl, cl
 je      .LBB2329_6
 jmp     .LBB2329_9
.LBB2329_3:
 mov     si, 16
 test    cl, cl
 jne     .LBB2329_9
.LBB2329_6:
 test    si, si
 je      .LBB2329_9
     pmovmskb edi, xmm0
     mov     ecx, esi
     and     cl, 15
     mov     ebx, -1
     shl     ebx, cl
     cmp     si, 16
     not     ebx
     cmovae  ebx, r11d
 && any_newline.to_bitmask() & (1u16.checked_shl(advance_by).map_or(u16::MAX, |c| c - 1))
 test    ebx, edi
 if !self.newline
 je      .LBB2329_8
 self.newline = true;
 mov     byte, ptr, [rdx, +, 73], 1
 mov     cl, 1
.LBB2329_9:
 if (advance_by as usize) < ELEMENTS {
 movzx   edi, si
 cmp     si, 16
 jb      .LBB2329_10
.LBB2329_11:
 self.offset += advance_by as usize;
 add     qword, ptr, [rdx, +, 64], rdi
 if self.found {
 add     r10, 16
 cmp     byte, ptr, [rdx, +, 72], 0
 je      .LBB2329_1
 jmp     .LBB2329_12
.LBB2329_8:
 if (advance_by as usize) < ELEMENTS {
 xor     ecx, ecx
 movzx   edi, si
 cmp     si, 16
 jae     .LBB2329_11
.LBB2329_10:
 self.found = true;
 mov     byte, ptr, [rdx, +, 72], 1
 self.offset += advance_by as usize;
 add     qword, ptr, [rdx, +, 64], rdi
.LBB2329_12:
 }
 mov     rcx, qword, ptr, [rdx, +, 64]
 mov     qword, ptr, [rax, +, 64], rcx
 movzx   ecx, byte, ptr, [rdx, +, 72]
 mov     byte, ptr, [rax, +, 72], cl
 mov     ecx, dword, ptr, [rdx, +, 73]
 mov     dword, ptr, [rax, +, 73], ecx
 movzx   ecx, word, ptr, [rdx, +, 77]
 mov     word, ptr, [rax, +, 77], cx
 movzx   ecx, byte, ptr, [rdx, +, 79]
 mov     byte, ptr, [rax, +, 79], cl
 movaps  xmm0, xmmword, ptr, [rdx]
 movaps  xmm1, xmmword, ptr, [rdx, +, 16]
 movaps  xmm2, xmmword, ptr, [rdx, +, 32]
 movaps  xmm3, xmmword, ptr, [rdx, +, 48]
 movaps  xmmword, ptr, [rax, +, 48], xmm3
 movaps  xmmword, ptr, [rax, +, 32], xmm2
 movaps  xmmword, ptr, [rax, +, 16], xmm1
 movaps  xmmword, ptr, [rax], xmm0
 pop     rbx
 pop     rdi
 pop     rsi
 ret
```

Reply:
I didn't read carefully, but it looks like the LLVM optimized the bounds check into an `and`+`je` (instead of the ideal `jbe`). Good enough. 👍

The memcpy at the end of that function is concerning though... Mutating `self` instead of returning a copy should fix that, right?

Reply:
The `and` is only performed once though.

Great catch about the copies at the end.
Mutating self does remove them.